sql/pgwire: fix encoding of int4 and bpchar in tuples

rafiss opened the pull request:
fixes #58069

This includes 3 commits:

### sql/pgwire: make PGTest pass against PostgresSQL

There are a few minor differences: dataTypeSize is different for tuples,
and PG does not show seconds offset for times if it is zero.

### sql/pgwire: fix binary encoding of ints in tuples

### sql/pgwire: fix text encoding of bpchar in tuples

### sql/pgwire: fix encoding of collated strings

Release note (bug fix): Integers inside of tuples were not being encoded
properly when using the binary format for retrieving data. This is now
fixed, and the proper integer width is reported.

Release note (bug fix): Blank-padded chars (e.g. CHAR(3)) were not being
encoded correctly when returning results to the client. Now they
correctly include blank-padding when appropriate.

Release note (bug fix): Collated strings were not encoded with the
proper type OID when sending results to the client if the OID was
for the `char` type. This is now fixed.

rafiss committed:
sql/pgwire: make PGTest pass against PostgresSQL

There are a few minor differences: dataTypeSize is different for tuples,
and PG does not show seconds offset for times if it is zero.

Release note: None

rafiss committed:
sql/pgwire: fix binary encoding of ints in tuples

Release note (bug fix): Integers inside of tuples were not being encoded
properly when using the binary format for retrieving data. This is now
fixed, and the proper integer width is reported.

rafiss committed:
sql/pgwire: fix text encoding of bpchar in tuples

Release note (bug fix): Blank-padded chars (e.g. CHAR(3)) were not being
encoded correctly when returning results to the client. Now they
correctly include blank-padding when appropriate.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61013)
<!-- Reviewable:end -->

otan commented on pkg/sql/pgwire/types.go:
```diff
@@ -422,7 +410,7 @@ func (b *writeBuffer) writeBinaryDatum(
+		b.writeLengthPrefixedString(tree.ResolveBlankPaddedChar(string(*v), t))
 
 	case *tree.DCollatedString:
 		b.writeLengthPrefixedString(v.Contents)
```
do we need to resolve blank padded chars for collated strings too?

rafiss replied:
yeah i think i should add it

rafiss committed:
sql/pgwire: fix encoding of collated strings

Release note (bug fix): Collated strings were not encoded with the
proper type OID when sending results to the client if the OID was
for the `char` type. This is now fixed.

otan commented on pkg/sql/pgwire/encoding_test.go:
```diff
@@ -184,6 +184,11 @@ func TestEncodings(t *testing.T) {
 				// Unsupported.
 				continue
+			case *tree.DCollatedString:
+				// Decoding collated strings is unsupported by this test. The encoded
```
treating collated strings differently to postgres has truly been a pain :(

rafiss replied:
yeah the gift that keeps giving

otan reviewed (approved):
nice. are we backporting this?

rafiss commented:
tftr!

bors r=otan

rafiss commented:
bors r=otan
